### PR TITLE
Coverage: support instrumenting one function at a time

### PR DESCRIPTION
**Pull request by smowton**
This factors the coverage driver to pull apart parsing configuration and generating a coverage driver structure from actually performing the coverage, and factors the whole-program version to use a single-function version that operates under the same rules. This sets us up for an easy transition to applying coverage instrumentation per-function in JBMC.

Note I don't actually use this in JBMC yet as it does not yet use remove-exceptions in per-function mode. This PR is for reviewing the coverage alterations; I will bring in remove-exceptions and coverage-instrumentation together with the upcoming symex-driven function loading PR.

### CODE REVIEW

File: src/goto-instrument/cover.cpp
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: This might not (and will not) be reached if the function_filter filters out __CPROVER prefixed functions. But it must be called after instrument_cover_goals() (or requires a fix to is_non_coverage_assert()).

File: src/goto-instrument/cover.h
Comment (peterschrammel) [inline src/goto-instrument/cover.h]: not sure why the param name is needed here

File: src/goto-instrument/cover.h
Comment (peterschrammel) [inline src/goto-instrument/cover.h:58]: What about wrapping everything into a covert class to keep the configuration as state? This would avoid exposing the implementation details to the user of the interface.

**Comment by smowton** [comment]
@peterschrammel updated fixing the issue you pointed out. Re: exposing implementation details, if I wrap it all into one big class I'll have to expose *that*, so I think exposure happens either way.

**Comment by smowton** [comment]
@tautschnig @peterschrammel this one is now ready. Peter, I addressed your comments above.

**Comment by smowton** [comment]
Pinging @thk123 @chrisr-diffblue @mgudemann this is ready for review

**Comment by smowton** [comment]
Made a test-gen bump: https://github.com/diffblue/test-gen/pull/1428

**Review by chrisr-diffblue — APPROVED** [review]
Looks reasonable to me.

File: src/goto-instrument/cover.cpp
Comment (thk123) [inline src/goto-instrument/cover.cpp:248]: Can we here swap to using range based for loop? Or perhaps even a `std::foreach` since the body is so simple

**Review by thk123 — APPROVED** [review]
Seems fine to me, though of course as I am contractually bound to say, you might consider adding some unit tests for this functionality :stuck_out_tongue:

File: src/goto-instrument/cover.cpp
Comment (smowton) [reply src/goto-instrument/cover.cpp:248]: I'm going to plead "moved code" plus slim benefit and leave this alone :)

Comment (smowton) [reply src/goto-instrument/cover.cpp]: I argue that the existing coverage tests will do this for us when this comes into use. Those tests will appear with the symex-driven lazy loading PR, which will be this code's first user and will introduce a variant of the cbmc-java test directory amongst others, running all tests with symex-driven lazy loading (and therefore per-function GOTO program passes) enabled.

**Comment by smowton** [comment]
@thk123 I argue that the existing coverage tests will do this for us when this comes into use. Those tests will appear with the symex-driven lazy loading PR, which will be this code's first user and will introduce a variant of the cbmc-java test directory amongst others, running all tests with symex-driven lazy loading (and therefore per-function GOTO program passes) enabled.

File: src/goto-instrument/cover.cpp
Comment (tautschnig) [inline src/goto-instrument/cover.cpp:177]: Why does this need to be a pointer?

File: src/goto-instrument/cover.cpp
Comment (tautschnig) [inline src/goto-instrument/cover.cpp]: Can't this be an optional instead?

File: src/goto-instrument/cover.cpp
Comment (smowton) [reply src/goto-instrument/cover.cpp:177]: I'd rather not implement move constructors for all the filter kinds -- simpler to let them not support copy/move and pass a handle to the config object instead. Unique-ptr really takes the pain out of managing its lifespan in any case.

Comment (tautschnig) [reply src/goto-instrument/cover.cpp:177]: It just seems a lot of machinery for a very simple problem where performance doesn't matter either. Why not just return `cover_configt` by value?

Comment (smowton) [reply src/goto-instrument/cover.cpp:177]: Same thing -- needing copyable and movable everything in the config object. So much simpler to just allocate one and pass it around.

Comment (tautschnig) [reply src/goto-instrument/cover.cpp:177]: Ok. So while I _would_ have done it differently, you actually _did it_, so you win :-)

Comment (smowton) [reply src/goto-instrument/cover.cpp:177]: Are you happy to approve in that case? This is still blocked on code-owner approval

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
